### PR TITLE
ceph: purge a down osd with a job

### PR DESCRIPTION
**Pull request by xenolinux**
If an OSD is down and needs to be removed from the cluster,
a job can be run that will remove the OSD deployment
and purge the OSD from ceph. If the OSD is still up,
the purge will be rejected.

To purge multiple OSDs at the same time, the OSD IDs can be
specified as a comma-separated list.

Signed-off-by: Servesha Dudhgaonkar <sdudhgao@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/daemon/ceph/osd/remove.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	logger.Infof("removing the OSD prepare job %q", prepareJobName)
+	if err := k8sutil.DeleteBatchJob(context.Clientset, clusterInfo.Namespace, prepareJobName, false); err != nil {
+		if !kerrors.IsNotFound(err) {
+			// Continue deleting the OSD prepare job even if the deployment fails to be deleted
+			logger.Errorf("failed to delete prepare job for OSD %s. %v", prepareJobName, err)
+		}
+	}
+
+	// If running on a PVC, delete the OSD PVC
+	if pvcName, ispresent := prepareJob.GetLabels()["ceph.rook.io/pvc"]; ispresent {
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:97]: in golang it's common to use `ok` for this check in the map
```suggestion
	if pvcName, ok := prepareJob.GetLabels()["ceph.rook.io/pvc"]; ok {
```

File: pkg/daemon/ceph/osd/remove.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	if err := k8sutil.DeleteDeployment(context.Clientset, clusterInfo.Namespace, deploymentName); err != nil {
+		if !kerrors.IsNotFound(err) {
+			// Continue purging the OSD even if the deployment fails to be deleted
+			logger.Errorf("failed to delete deployment for OSD %d. %v", osdID, err)
+		}
+	}
+
+	// If running on a PVC, delete the OSD prepare job
+	prepareJobName := fmt.Sprintf("rook-ceph-osd-prepare-example-deviceset-0-#{osdID}")
+	prepareJob, err := context.Clientset.BatchV1().Jobs(clusterInfo.Namespace).Get(prepareJobName, metav1.GetOptions{})
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:87]: Need to check if there was an err after retrieving the job. If it's not found, we can skip this step.

File: pkg/daemon/ceph/osd/remove.go (hunk 1 above)
Comment (travisn) [reply pkg/daemon/ceph/osd/remove.go:97]: The `prepareJob` might not have been found on line 87, so we need to skip this check if it's `nil`.

File: pkg/daemon/ceph/osd/remove.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+func removeOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdID int) error {
+
+	// Check if the OSD is down
+	osdDump, err := client.GetOSDDump(context, clusterInfo)
+	if err != nil {
+		return errors.Wrap(err, "failed to get osd dump")
+	}
+
+	for _, osdStatus := range osdDump.OSDs {
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:64]: This `for` loop should also move up to the other method so we just compare the OSD IDs that are requested for removal. This looks like it loops through all OSDs.

File: pkg/daemon/ceph/osd/remove.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+			return err
+		}
+
+		const upStatus int64 = 1
+		var osdDown bool
+		if status == upStatus {
+			logger.Debugf("osd.%d is healthy.", id)
+			osdDown = false
+			continue
+
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:84]: nit: remove the blank line before a closing bracket

File: pkg/daemon/ceph/osd/remove.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+		}
+
+		logger.Debugf("osd.%d is marked 'DOWN'", id)
+
+		// Get the host where the OSD is found
+		hostName, err := client.GetCrushHostName(context, clusterInfo, osdID)
+
+		if !osdDown {
+			return fmt.Errorf("OSD %d cannot be removed unless it is 'down'", osdID)
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:93]: No need for the `osdDown` var if you directly return this error above on line 83 in place of the continue.

File: pkg/daemon/ceph/osd/remove.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+		logger.Infof("removing the OSD deployment %q", deploymentName)
+		if err := k8sutil.DeleteDeployment(context.Clientset, clusterInfo.Namespace, deploymentName); err != nil {
+			if !kerrors.IsNotFound(err) {
+				// Continue purging the OSD even if the deployment fails to be deleted
+				logger.Errorf("failed to delete deployment for OSD %d. %v", osdID, err)
+			}
+		}
+
+		// If running on a PVC, delete the OSD prepare job
+		prepareJobName := fmt.Sprintf("rook-ceph-osd-prepare-example-deviceset-0-#{osdID}")
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:114]: The osdID is not part of the osd prepare job name. We will need to lookup the osd prepare job differently, such as seeing that it has the same PVC as the OSD deployment that is being deleted.

File: pkg/daemon/ceph/osd/remove.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+				if err := context.Clientset.CoreV1().PersistentVolumeClaims(clusterInfo.Namespace).Delete(pvcName, &metav1.DeleteOptions{}); err != nil {
+					if !kerrors.IsNotFound(err) {
+						// Continue deleting the OSD PVC even if PVC deletion fails
+						logger.Errorf("failed to delete pvc for OSD %s. %v", pvcName, err)
+					}
+				}
+			}
+
+			// purge the osd
+			output, err = context.Executor.ExecuteCommandWithOutput("ceph", "osd", "purge", fmt.Sprintf("osd.%d", osdID), "--force", "--yes-i-really-mean-it")
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:137]: From this point on should not be in the *else* statement. We still want to purge the OSD even if the osd prepare job was not found.

File: pkg/daemon/ceph/osd/remove.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+	}
+
+	for _, osdStatus := range osdDump.OSDs {
+		id64, err := osdStatus.OSD.Int64()
+		if err != nil {
+			continue
+		}
+		id := int(id64)
+
+		logger.Debugf("validating status of osd.%d", id)
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:52]: We should only be validating the status of the OSDs that are in the list of OSDs to remove. And if it's in the list, it can be removed as part of this for loop instead of having a separate for loop.

File: pkg/daemon/ceph/osd/remove.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+	for _, osdIDStr := range osdsToRemove {
+		logger.Infof("removing OSD %q", osdIDStr)
+		osdID, err := strconv.Atoi(osdIDStr)
+		if err != nil {
+			logger.Errorf("invalid OSD ID: %s. %v", osdIDStr, err)
+			continue
+		}
+
+		if err := removeOSD(context, clusterInfo, osdID); err != nil {
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:75]: The OSDs will be removed even if they are not down. `removeOSD()` should only be called if the osd is down.

File: pkg/daemon/ceph/osd/remove.go [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+	return nil
+}
+
+func removeOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdID int) error {
+
+	// Get the host where the OSD is found
+	hostName, err := client.GetCrushHostName(context, clusterInfo, osdID)
+
+	// Mark the OSD as out. If it is not down, it is expected this command will fail, in which case
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:88]: An OSD can still be marked as `out` even if it is not down, so I don't expect this to fail for that reason.

File: pkg/daemon/ceph/osd/remove.go [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+	}
+	logger.Infof("removing the OSD deployment %q", deploymentName)
+	if err := k8sutil.DeleteDeployment(context.Clientset, clusterInfo.Namespace, deploymentName); err != nil {
+		if err != nil {
+			// Continue purging the OSD even if the deployment fails to be deleted
+			logger.Errorf("failed to delete deployment for OSD %d. %v", osdID, err)
+		}
+	}
+	if pvcName, ok := deployment.GetLabels()["ceph.rook.io/pvc"]; ok {
+		prepareJobList, err := context.Clientset.BatchV1().Jobs(clusterInfo.Namespace).List(metav1.ListOptions{LabelSelector: "ceph.rook.io/pvc=" + pvcName})
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:109]: The osd prepare job has the same label as the osd daemon? I didn't realize that was already there.

File: pkg/daemon/ceph/osd/remove.go [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+
+		if err := removeOSD(context, clusterInfo, osdID); err != nil {
+			logger.Errorf("failed to remove OSD %d. %v", osdID, err)
+		}
+	}
+
+	return nil
+}
+
+func removeOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdID int) error {
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:83]: When this func is called, the decision should already be made that we are going to do everything to clean up the OSD. Removing all the resources is a best effort, meaning that we should try to continue removing the OSD even if some of the resources aren't found. If resources are not found, we just log it and continue with the next command.

File: pkg/daemon/ceph/osd/remove.go [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+				// Continue deleting the OSD PVC even if PVC deletion fails
+				logger.Errorf("failed to delete pvc for OSD %s. %v", pvcName, err)
+			}
+		}
+	}
+
+	// purge the osd
+	output, err = context.Executor.ExecuteCommandWithOutput("ceph", "osd", "purge", fmt.Sprintf("osd.%d", osdID), "--force", "--yes-i-really-mean-it")
+	if err != nil {
+		return errors.Wrapf(err, fmt.Sprintf("failed ceph status. %s", output))
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:136]: log instead of return

File: pkg/daemon/ceph/osd/remove.go (hunk 11 above)
Comment (xenolinux) [reply pkg/daemon/ceph/osd/remove.go:109]: Yes. osd prepare job and osd deployment/pod have same label.

File: pkg/daemon/ceph/osd/remove.go (hunk 10 above)
Comment (xenolinux) [reply pkg/daemon/ceph/osd/remove.go:88]: Right, will fix it.

File: pkg/daemon/ceph/osd/remove.go (hunk 13 above)
Comment (xenolinux) [reply pkg/daemon/ceph/osd/remove.go:136]: ack

File: pkg/daemon/ceph/osd/remove.go [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+		}
+		logger.Debugf("validating status of osd.%d", osdID)
+		status, _, err := osdDump.StatusByID(int64(osdID))
+		if err != nil {
+			return errors.Wrap(err, fmt.Sprintf("failed to get osd status for osd %v", osdID))
+
+		}
+		const upStatus int64 = 1
+		if status == upStatus {
+			logger.Debugf("osd.%d is healthy.It cannot be removed unless it is 'down'", osdID)
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:59]:
```suggestion
			logger.Debugf("osd.%d is healthy. It cannot be removed unless it is 'down'", osdID)
```

File: pkg/daemon/ceph/osd/remove.go [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+		log.Errorf(fmt.Sprintf("failed to execute command 'ceph osd out' for. %s: %v", output, err))
+	}
+
+	// Remove the OSD deployment
+	deploymentName := fmt.Sprintf("rook-ceph-osd-%d", osdID)
+	deployment, err := context.Clientset.AppsV1().Deployments(clusterInfo.Namespace).Get(deploymentName, metav1.GetOptions{})
+	if err != nil {
+		logger.Errorf(fmt.Sprintf("error fetching the deployment %s : %v", deploymentName, err))
+	}
+	logger.Infof("removing the OSD deployment %q", deploymentName)
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:85]: I'd suggest putting lines 85-117 inside an `else` statement. Then you don't need to check for `deployment != nil` on line 92, and it's clear what code will be skipped if line 81 has an error.

Comment (xenolinux) [reply pkg/daemon/ceph/osd/remove.go:85]: Okay. Will do

**Comment by xenolinux** [comment]
Test result: The code is working fine. 
Logs of rook-ceph-purge-osd job https://docs.google.com/document/d/1yr9nvkjuIt0dsEGPZQ0EH8ItPHk2XT6I5bHyJjCJ5vA/edit?usp=sharing

**Comment by xenolinux** [comment]
Test result: The code is working fine. 
Logs of rook-ceph-purge-osd job https://docs.google.com/document/d/1yr9nvkjuIt0dsEGPZQ0EH8ItPHk2XT6I5bHyJjCJ5vA/edit?usp=sharing

File: cluster/examples/kubernetes/ceph/osd-purge.yaml [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+            - name: ROOK_LOG_LEVEL
+              value: DEBUG
+          volumeMounts:
+            - mountPath: /etc/ceph
+              name: ceph-conf-emptydir
+      volumes:
+        - emptyDir: {}
+          name: ceph-conf-emptydir
+      restartPolicy: Never
+
```
Comment (leseb) [inline cluster/examples/kubernetes/ceph/osd-purge.yaml:56]: Unnecessary empty lines

File: pkg/daemon/ceph/osd/remove.go [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+func removeOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdID int) {
+	// Get the host where the OSD is found
+	hostName, err := client.GetCrushHostName(context, clusterInfo, osdID)
+
+	// Mark the OSD as out.
+	output, err := context.Executor.ExecuteCommandWithOutput("ceph", "osd", "out", fmt.Sprintf("osd.%d", osdID))
+	if err != nil {
+		log.Errorf(fmt.Sprintf("failed to execute command 'ceph osd out' for. %s: %v", output, err))
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:76]: Why are we using this logger all of a sudden? Just use the current  `logger` and then remove the mod dependency.

File: pkg/daemon/ceph/osd/remove.go [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+	for _, osdIDStr := range osdsToRemove {
+		osdID, err := strconv.Atoi(osdIDStr)
+		if err != nil {
+			logger.Errorf("invalid OSD ID: %s. %v", osdIDStr, err)
+			continue
+		}
+		logger.Debugf("validating status of osd.%d", osdID)
+		status, _, err := osdDump.StatusByID(int64(osdID))
+		if err != nil {
+			return errors.Wrap(err, fmt.Sprintf("failed to get osd status for osd %v", osdID))
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:54]: Use `wrapf()` instead.

File: pkg/daemon/ceph/osd/remove.go [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+
+	return nil
+}
+
+func removeOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdID int) {
+	// Get the host where the OSD is found
+	hostName, err := client.GetCrushHostName(context, clusterInfo, osdID)
+
+	// Mark the OSD as out.
+	output, err := context.Executor.ExecuteCommandWithOutput("ceph", "osd", "out", fmt.Sprintf("osd.%d", osdID))
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:74]: Use `NewCephCommand()` instead of `context.Executor.ExecuteCommandWithOutput()`. The latter is only needed when not executing Ceph commands.

File: pkg/daemon/ceph/osd/remove.go (hunk 17 above)
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:76]: You don't need `fmt.Sprintf()` if you use `log.Errorf()`.

File: pkg/daemon/ceph/osd/remove.go [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+	if err != nil {
+		log.Errorf(fmt.Sprintf("failed to execute command 'ceph osd out' for. %s: %v", output, err))
+	}
+
+	// Remove the OSD deployment
+	deploymentName := fmt.Sprintf("rook-ceph-osd-%d", osdID)
+	deployment, err := context.Clientset.AppsV1().Deployments(clusterInfo.Namespace).Get(deploymentName, metav1.GetOptions{})
+	if err != nil {
+		logger.Errorf(fmt.Sprintf("error fetching the deployment %s : %v", deploymentName, err))
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:83]: ditto

File: pkg/daemon/ceph/osd/remove.go [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+		if err := k8sutil.DeleteDeployment(context.Clientset, clusterInfo.Namespace, deploymentName); err != nil {
+			if err != nil {
+				// Continue purging the OSD even if the deployment fails to be deleted
+				logger.Errorf("failed to delete deployment for OSD %d. %v", osdID, err)
+			}
+		}
+		if pvcName, ok := deployment.GetLabels()["ceph.rook.io/pvc"]; ok {
+			prepareJobList, err := context.Clientset.BatchV1().Jobs(clusterInfo.Namespace).List(metav1.ListOptions{LabelSelector: "ceph.rook.io/pvc=" + pvcName})
+			if err != nil && !kerrors.IsNotFound(err) {
+				logger.Errorf(fmt.Sprintf("failed to list prepareJobs with labels %s: %v ", "ceph.rook.io/pvc="+pvcName, err))
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:95]: ditto

File: pkg/daemon/ceph/osd/remove.go [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+			}
+			// Remove osd prepare job
+			for _, prepareJob := range prepareJobList.Items {
+				logger.Infof("removing the OSD prepare job %q", prepareJob.GetName())
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:99]:
```suggestion
				logger.Infof("removing the osd prepare job %q", prepareJob.GetName())
```

File: pkg/daemon/ceph/osd/remove.go [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+				if err := k8sutil.DeleteBatchJob(context.Clientset, clusterInfo.Namespace, prepareJob.GetName(), false); err != nil {
+					if err != nil {
+						// Continue deleting the OSD prepare job even if the deployment fails to be deleted
+						logger.Errorf("failed to delete prepare job for OSD %s. %v", prepareJob.GetName(), err)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:103]:
```suggestion
						logger.Errorf("failed to delete prepare job for osd %q. %v", prepareJob.GetName(), err)
```

File: pkg/daemon/ceph/osd/remove.go [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+					}
+				}
+
+				logger.Infof("removing the OSD PVC %q", pvcName)
+				if err := context.Clientset.CoreV1().PersistentVolumeClaims(clusterInfo.Namespace).Delete(pvcName, &metav1.DeleteOptions{}); err != nil {
+					if err != nil {
+						// Continue deleting the OSD PVC even if PVC deletion fails
+						logger.Errorf("failed to delete pvc for OSD %s. %v", pvcName, err)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:111]:
```suggestion
						logger.Errorf("failed to delete pvc for OSD %q. %v", pvcName, err)
```

File: pkg/daemon/ceph/osd/remove.go [hunk 25]
```diff
@@ -0,0 +1,131 @@
...
+					}
+				}
+			}
+		}
+	}
+
+	// purge the osd
+	output, err = context.Executor.ExecuteCommandWithOutput("ceph", "osd", "purge", fmt.Sprintf("osd.%d", osdID), "--force", "--yes-i-really-mean-it")
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:119]: ditto

File: pkg/daemon/ceph/osd/remove.go [hunk 26]
```diff
@@ -0,0 +1,131 @@
...
+	if err != nil {
+		logger.Errorf("failed ceph status. %s", output)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:121]: Wrong message

File: pkg/daemon/ceph/osd/remove.go [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
+	}
+
+	// Attempting to remove the parent host. Errors can be ignored if there are other OSDs on the same host
+	output, err = context.Executor.ExecuteCommandWithOutput("ceph", "osd", "crush", "rm", hostName)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:125]: ditto

File: pkg/daemon/ceph/osd/remove.go [hunk 28]
```diff
@@ -0,0 +1,131 @@
...
+	if err != nil {
+		logger.Infof("unable to remove CRUSH host %q. %v", hostName, err)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:127]: Errorf

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/daemon/ceph/osd/remove.go (hunk 17 above)
Comment (xenolinux) [reply pkg/daemon/ceph/osd/remove.go:76]: ack

File: pkg/daemon/ceph/osd/remove.go (hunk 18 above)
Comment (xenolinux) [reply pkg/daemon/ceph/osd/remove.go:54]: ack

File: pkg/daemon/ceph/osd/remove.go [hunk 29]
```diff
@@ -0,0 +1,133 @@
...
+	for _, osdIDStr := range osdsToRemove {
+		osdID, err := strconv.Atoi(osdIDStr)
+		if err != nil {
+			logger.Errorf("invalid OSD ID: %s. %v", osdIDStr, err)
+			continue
+		}
+		logger.Debugf("validating status of osd.%d", osdID)
+		status, _, err := osdDump.StatusByID(int64(osdID))
+		if err != nil {
+			return errors.Wrapf(err, "failed to get osd status for osd %v", osdID)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:53]:
```suggestion
			return errors.Wrapf(err, "failed to get osd status for osd %d", osdID)
```

File: pkg/daemon/ceph/osd/remove.go [hunk 30]
```diff
@@ -0,0 +1,133 @@
...
+
+func removeOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdID int) {
+	// Get the host where the OSD is found
+	hostName, err := client.GetCrushHostName(context, clusterInfo, osdID)
+
+	// Mark the OSD as out.
+	args := []string{"osd", "out", fmt.Sprintf("osd.%d", osdID)}
+	_, err = client.NewCephCommand(context, clusterInfo, args).Run()
+	if err != nil {
+		logger.Errorf("failed to execute command 'ceph osd out' for. %v", err)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:76]: for ?

File: pkg/daemon/ceph/osd/remove.go [hunk 31]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	// Remove the OSD deployment
+	deploymentName := fmt.Sprintf("rook-ceph-osd-%d", osdID)
+	deployment, err := context.Clientset.AppsV1().Deployments(clusterInfo.Namespace).Get(deploymentName, metav1.GetOptions{})
+	if err != nil {
+		logger.Errorf("error fetching the deployment %s : %v", deploymentName, err)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:83]:
```suggestion
		logger.Errorf("failed to fetch the deployment %q. %v", deploymentName, err)
```

File: pkg/daemon/ceph/osd/remove.go [hunk 32]
```diff
@@ -0,0 +1,133 @@
...
+		if err := k8sutil.DeleteDeployment(context.Clientset, clusterInfo.Namespace, deploymentName); err != nil {
+			if err != nil {
+				// Continue purging the OSD even if the deployment fails to be deleted
+				logger.Errorf("failed to delete deployment for OSD %d. %v", osdID, err)
+			}
+		}
+		if pvcName, ok := deployment.GetLabels()["ceph.rook.io/pvc"]; ok {
+			prepareJobList, err := context.Clientset.BatchV1().Jobs(clusterInfo.Namespace).List(metav1.ListOptions{LabelSelector: "ceph.rook.io/pvc=" + pvcName})
+			if err != nil && !kerrors.IsNotFound(err) {
+				logger.Errorf("failed to list prepareJobs with labels %s: %v ", "ceph.rook.io/pvc="+pvcName, err)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:95]:
```suggestion
				logger.Errorf("failed to list prepareJobs with labels %q. %v ", "ceph.rook.io/pvc="+pvcName, err)
```

**Review by leseb — CHANGES_REQUESTED** [review]

File: go.sum [hunk 33]
```diff
@@ -43,8 +43,10 @@ github.com/StackExchange/wmi v0.0.0-20180116203802-5d049714c4a6/go.mod h1:3eOhrU
 github.com/agnivade/levenshtein v1.0.1/go.mod h1:CURSv5d9Uaml+FovSIICkLbAUZ9S4RqaHDIsdSBg7lM=
 github.com/airbrake/gobrake v3.6.1+incompatible/go.mod h1:wM4gu3Cn0W0K7GUuVWnlXZU11AGBXMILnrdOU8Kn00o=
 github.com/alecthomas/template v0.0.0-20160405071501-a0175ee3bccc/go.mod h1:LOuyumcjzFXgccqObfd/Ljyb9UuFJ6TxHnclSeseNhc=
+github.com/alecthomas/template v0.0.0-20190718012654-fb15b899a751 h1:JYp7IbQjafoB+tBA3gMyHYHrpOtNuDiK/uB5uXxq5wM=
```
Comment (leseb) [inline go.sum:46]: Why did you change this?

**Review by leseb — CHANGES_REQUESTED** [review]
lgtm, the yaml comes as an example but is not used by Rook AFAIK. Also, I think we need to remove the go.sum changes, not sure why they are here.

File: go.sum (hunk 33 above)
Comment (xenolinux) [reply go.sum]: Removed go.sum file from the patch

File: pkg/daemon/ceph/osd/remove.go [hunk 34]
```diff
@@ -0,0 +1,133 @@
...
+				}
+			}
+		}
+	}
+
+	// purge the osd
+	purgeosdargs := []string{"osd", "purge", fmt.Sprintf("osd.%d", osdID), "--force", "--yes-i-really-mean-it"}
+	_, err = client.NewCephCommand(context, clusterInfo, purgeosdargs).Run()
+	if err != nil {
+		logger.Errorf("failed to purge the osd. %v", err)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:122]:
```suggestion
		logger.Errorf("failed to purge osd.%d. %v", osdID, err)
```

File: pkg/daemon/ceph/osd/remove.go [hunk 35]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	// Attempting to remove the parent host. Errors can be ignored if there are other OSDs on the same host
+	hostargs := []string{"osd", "crush", "rm", hostName}
+	_, err = client.NewCephCommand(context, clusterInfo, hostargs).Run()
+	if err != nil {
+		logger.Errorf("unable to remove CRUSH host %q. %v", hostName, err)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:129]:
```suggestion
		logger.Errorf("failed to remove CRUSH host %q. %v", hostName, err)
```

File: pkg/daemon/ceph/osd/remove.go [hunk 36]
```diff
@@ -0,0 +1,133 @@
...
+		logger.Debugf("osd.%d is marked 'DOWN'. Removing it", osdID)
+		removeOSD(context, clusterInfo, osdID)
+	}
+
+	return nil
+}
+
+func removeOSD(context *clusterd.Context, clusterInfo *client.ClusterInfo, osdID int) {
+	// Get the host where the OSD is found
+	hostName, err := client.GetCrushHostName(context, clusterInfo, osdID)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:70]: You need to handle the error.

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/daemon/ceph/osd/remove.go (hunk 36 above)
Comment (xenolinux) [reply pkg/daemon/ceph/osd/remove.go:70]: ack

File: pkg/daemon/ceph/osd/remove.go [hunk 37]
```diff
@@ -0,0 +1,136 @@
...
+	if err != nil {
+		logger.Errorf("failed to get the host where osd is found %v", err)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:72]:
```suggestion
		logger.Errorf("failed to get the host where osd.%d is running. %v", osdID , err)
```

File: pkg/daemon/ceph/osd/remove.go [hunk 38]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	// Mark the OSD as out.
+	args := []string{"osd", "out", fmt.Sprintf("osd.%d", osdID)}
+	_, err = client.NewCephCommand(context, clusterInfo, args).Run()
+	if err != nil {
+		logger.Errorf("failed to execute command 'ceph osd out' %v", err)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:79]:
```suggestion
		logger.Errorf("failed to execute command 'ceph osd out'. %v", err)
```

**Review by leseb — CHANGES_REQUESTED** [review]
almost there

File: cmd/rook/ceph/osd.go [hunk 39]
```diff
@@ -240,6 +252,29 @@ func prepareOSD(cmd *cobra.Command, args []string) error {
...
+	required = []string{"mon-endpoints", "ceph-username", "ceph-secret"}
+	if err := flags.VerifyRequiredFlags(osdCmd, required); err != nil {
+		return err
+	}
+
+	commonOSDInit(osdRemoveCmd)
+
+	context := createContext()
+
+	// Run OSD start sequence
```
Comment (leseb) [inline cmd/rook/ceph/osd.go:270]:
```suggestion
	// Run OSD remove sequence
```

File: pkg/daemon/ceph/osd/remove.go [hunk 40]
```diff
@@ -0,0 +1,136 @@
...
+	hostName, err := client.GetCrushHostName(context, clusterInfo, osdID)
+	if err != nil {
+		logger.Errorf("failed to get the host where osd.%d is running. %v", osdID, err)
+	}
+
+	// Mark the OSD as out.
+	args := []string{"osd", "out", fmt.Sprintf("osd.%d", osdID)}
+	_, err = client.NewCephCommand(context, clusterInfo, args).Run()
+	if err != nil {
+		logger.Errorf("failed to execute command 'ceph osd out'. %v", err)
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:79]:
```suggestion
		logger.Errorf("failed to exclude osd.%d out of the crush map. %v", osdID, err)
```

File: pkg/daemon/ceph/osd/remove.go [hunk 41]
```diff
@@ -0,0 +1,136 @@
...
+		logger.Errorf("failed to fetch the deployment %q. %v", deploymentName, err)
+	} else {
+		logger.Infof("removing the OSD deployment %q", deploymentName)
+		if err := k8sutil.DeleteDeployment(context.Clientset, clusterInfo.Namespace, deploymentName); err != nil {
+			if err != nil {
+				// Continue purging the OSD even if the deployment fails to be deleted
+				logger.Errorf("failed to delete deployment for OSD %d. %v", osdID, err)
+			}
+		}
+		if pvcName, ok := deployment.GetLabels()["ceph.rook.io/pvc"]; ok {
```
Comment (leseb) [inline pkg/daemon/ceph/osd/remove.go:95]: Please make  `ceph.rook.io/pvc` a constant so you use it multiple times.

**Review by leseb — CHANGES_REQUESTED** [review]

**Review by leseb — APPROVED** [review]

**Comment by leseb** [comment]
@xenolinux I just rebased your PR to latest master.

File: pkg/daemon/ceph/osd/remove.go [hunk 42]
```diff
@@ -0,0 +1,137 @@
...
+		return errors.Wrap(err, "failed to get osd dump")
+	}
+
+	for _, osdIDStr := range osdsToRemove {
+		osdID, err := strconv.Atoi(osdIDStr)
+		if err != nil {
+			logger.Errorf("invalid OSD ID: %s. %v", osdIDStr, err)
+			continue
+		}
+		logger.Debugf("validating status of osd.%d", osdID)
```
Comment (travisn) [inline pkg/daemon/ceph/osd/remove.go:50]: Let's call `logger.Infof()` instead of `Debugf()` everywhere in this `remove.go`. These are all helpful messages for troubleshooting in case an osd is removed or not.

Comment (xenolinux) [reply pkg/daemon/ceph/osd/remove.go:50]: ack. Will do
